Mathlib/LinearAlgebra/Dimension/RankNullity.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.LinearAlgebra.Dimension.Constructions
+
+/-!
+
+# The rank nullity theorem
+
+In this file we provide the rank nullity theorem as a typeclass, and prove various corollaries
+of the theorem. The main definition is `HasRankNullity.{w} R`, which states that
+1. Every `R`-module `M : Type w` has a linear independent subset of cardinality `Module.rank R M`.
+2. `rank (M ⧸ N) + rank N = rank M` for every `R`-module `M : Type w` and every `N : Submodule R M`.
+
+The following instances are provided in mathlib:
+1. `DivisionRing.hasRankNullity` for division rings in `LinearAlgebra/Dimension/DivisionRing.lean`.
+2. `IsDomain.hasRankNullity` for commutative domains in `LinearAlgebra/Dimension/Localization.lean`.
+
+TODO: prove the rank-nullity theorem for `[Ring R] [IsDomain R] [StrongRankCondition R]`.
```

Comment:
Are we sure it holds?

Reply:
I'm fairly certain that it holds. A domain that is not Ore is of infinite rank (see Cohn's *Skew Field* Proposition 1.3.6) and thus does not satisfy strong rank condition. Hence domains with strong rank condition are Ore and we can mimic the proof here.

Reply:
Nice! Since this is probably not very well known, do you mind add the reference?